src/compiler/chainrules.jl

```diff
@@ -11,12 +11,17 @@ such that if a suitable rule is defined later, the generated function will recom
 """
 function has_chain_rrule(T)
   m = meta(Tuple{typeof(rrule),T.parameters...})
-  if m.method !== chainrules_fallback
+  if m.method === chainrules_fallback
+    # no rule exists
+    return false, m.instance
+  elseif Core.Compiler.return_type(rrule, Tuple{T.parameters...}) === Nothing
+    Core.println("  has_chain_rrule got Nothing")
+    # or we hit a rule telling us to keep digging
```

Comment:
Is excluding this likely to affect compile time? IIUC @mcabbott 's reason for keeping this branch was to avoid having to do type inference some of the time.

Reply:
Yes, I was initially thinking not to call `Compiler.return_type` more times than needed, but without thinking much about whether that actually matters. This does seem to produce specialisations, so I guess there may be some cost, even if all of them are very simple?
```
julia> f(x) = nothing; f(1); f(1.0); f(1+im);

julia> first(methods(f)).specializations
svec(MethodInstance for f(::Int64), MethodInstance for f(::Float64), MethodInstance for f(::Complex{Int64}), #undef, #undef, #undef, #undef, #undef)

julia> g(x) = nothing; Core.Compiler.return_type(g, Tuple{Int}); Core.Compiler.return_type(g, Tuple{Float64}); Core.Compiler.return_type(g, Tuple{String});

julia> first(methods(g)).specializations
svec(MethodInstance for g(::Int64), MethodInstance for g(::Float64), MethodInstance for g(::String), #undef, #undef, #undef, #undef, #undef)
```

Reply:
I am not sure that it will actually make a difference in this case though, I think we might already have had the specialization made to the the `m.instance` made.

Reply:
IRTools I don't know. 

Is there some subtlety of testing the negative `if m.method !== chainrules_fallback` (before this PR) which I'm overlooking?

Reply:
https://github.com/FluxML/Zygote.jl/pull/967/commits/fbb3fcf09da7bd1773612c501f4fb0e6e5d34a6f makes a change like the suggestion. See what CI thinks of it.